groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+{
+    static_assert(
+        BloombergLP::bslstl::
+                Variant_HasUniqueType<TYPE, variant<HEAD, TAIL...> >::value ==
+            true,
+        "Type is not a unique alternative in the variant");
+
+    reset();
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<TYPE *> BSLS_UTIL_ADDRESSOF(d_union),
+        AllocBase::mechanism(),
+        il,
+        std::forward<ARGS>(args)...);
+    d_type = BloombergLP::bslstl::
+        Variant_TypeToIndexImpl<0, TYPE, HEAD, TAIL...>::value;
+
+    return bsl::get<
+        BloombergLP::bslstl::Variant_TypeToIndex<TYPE, variant>::value>(*this);
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX, class... ARGS>
+typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type&
+variant<HEAD, TAIL...>::emplace(ARGS&&... args)
+{
+    typedef typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type
+        TYPE;
+    emplace<TYPE>(std::forward<ARGS>(args)...);
+    return bsl::get<INDEX>(*this);
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX, class INIT_LIST_TYPE, class... ARGS>
+typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type&
+variant<HEAD, TAIL...>::emplace(std::initializer_list<INIT_LIST_TYPE> il,
+                                ARGS&&...                             args)
+{
+    typedef typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type
+        TYPE;
+    emplace<TYPE>(il, std::forward<ARGS>(args)...);
+    return bsl::get<INDEX>(*this);
+}
+#else   // BSL_VARIANT_FULL_IMPLEMENTATIO
+
+template <class HEAD, class... TAIL>
+template <class TYPE>
+TYPE& variant<HEAD, TAIL...>::emplace()
+{
+    reset();
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<TYPE *> BSLS_UTIL_ADDRESSOF(d_union),
+        AllocBase::mechanism());
+    d_type = BloombergLP::bslstl::
+        Variant_TypeToIndexImpl<0, TYPE, HEAD, TAIL...>::value;
+
+    return bsl::get<
+        BloombergLP::bslstl::Variant_TypeToIndex<TYPE, variant>::value>(*this);
+}
+
+template <class HEAD, class... TAIL>
+template <class TYPE, class ARG>
+TYPE& variant<HEAD, TAIL...>::emplace(
+                                    BSLS_COMPILERFEATURES_FORWARD_REF(ARG) arg)
+{
+    reset();
+    BloombergLP::bslma::ConstructionUtil::construct(
+        reinterpret_cast<TYPE *> BSLS_UTIL_ADDRESSOF(d_union),
+        AllocBase::mechanism(),
+        BSLS_COMPILERFEATURES_FORWARD(ARG, arg));
+    d_type = BloombergLP::bslstl::
+        Variant_TypeToIndexImpl<0, TYPE, HEAD, TAIL...>::value;
+
+    return bsl::get<
+        BloombergLP::bslstl::Variant_TypeToIndex<TYPE, variant>::value>(*this);
+}
+
+template <class HEAD, class... TAIL>
+template <size_t INDEX>
+typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type&
+variant<HEAD, TAIL...>::emplace()
+{
+    typedef typename variant_alternative<INDEX, variant<HEAD, TAIL...> >::type
+        TYPE;
+    emplace<TYPE>();
```

Comment:
Good catch. That's unfortunate, I'll modify it so it emplaces based on the index

Reply:
This is outdated code. The TYPE emplace forwards to the INDEX emplace now. I don't know git enough to force this to resync.  Remind me of which `reinterpret_cast` problem you are referring to.

Reply:
Right, it seems you did change it so that it calls the `INDEX` version and I must have been looking at outdated code when I made my last comment, so please ignore me